In [1]:
import joblib
import pandas as pd
from sklearn.model_selection import StratifiedKFold, cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
import time

# Load feature-engineered data
X_train = joblib.load('data/X_train_fe.pkl')
y_train = joblib.load('data/y_train.pkl')

In [2]:
models = {
    'Logistic Regression': LogisticRegression(class_weight='balanced', max_iter=1000, random_state=42),
    'Random Forest': RandomForestClassifier(class_weight='balanced', n_estimators=200, random_state=42),
    'XGBoost': XGBClassifier(
        scale_pos_weight=(y_train.value_counts()[0] / y_train.value_counts()[1]),
        eval_metric='logloss',
        random_state=42
    )
}

In [3]:
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

results = []

for name, model in models.items():
    start = time.time()
    f1_scores = cross_val_score(model, X_train, y_train, cv=cv, scoring='f1')
    roc_auc_scores = cross_val_score(model, X_train, y_train, cv=cv, scoring='roc_auc')
    elapsed = time.time() - start
    
    results.append({
        'Model': name,
        'F1 Mean': f1_scores.mean(),
        'F1 Std': f1_scores.std(),
        'ROC-AUC Mean': roc_auc_scores.mean(),
        'ROC-AUC Std': roc_auc_scores.std(),
        'Time (s)': round(elapsed, 2)
    })

results_df = pd.DataFrame(results).sort_values('F1 Mean', ascending=False)
results_df

f:\SLIIT HND Y2S1\foundation to AI\FAI_Project\venv\Lib\site-packages\sklearn\linear_model\_logistic.py:599: ConvergenceWarning: lbfgs failed to converge after 1000 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=1000).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
f:\SLIIT HND Y2S1\foundation to AI\FAI_Project\venv\Lib\site-packages\sklearn\linear_model\_logistic.py:599: ConvergenceWarning: lbfgs failed to converge after 1000 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=1000).
You might also want to scale the data as shown in:
    https://scikit-

,Model,F1 Mean,F1 Std,ROC-AUC Mean,ROC-AUC Std,Time (s)
1,Random Forest,0.580346,0.011172,0.966303,0.001983,54.59
2,XGBoost,0.567024,0.018858,0.965251,0.002857,16.23
0,Logistic Regression,0.457462,0.007449,0.933733,0.002285,18.95


Since Random forest has a F1 Mean of 0.580346 which is the highest, from the three, I will chose this model to move forward